[c10d] separate the codes for GPU stream synchronization and CPU thread synchronization

shuqiangzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #137295

code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.

In a longer term, we would like to give user the flexibility of bypassing the watchdog thread and handle the collective error by themselves.

Test Plan:
python test/distributed/test_c10d_nccl.py NcclErrorHandlingTest


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

shuqiangzhang committed:
[c10d] separte the concepts of stream synchronization and CPU blocking
code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.
Test Plan:
Existing UT passed
Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] separte the concepts of stream synchronization and CPU blocking"


code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.
Test Plan:
Existing UT passed
Reviewers:

Subscribers:

Tasks:

Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] separte the concepts of stream synchronization and CPU blocking"


code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.
Test Plan:
Existing UT passed
Reviewers:

Subscribers:

Tasks:

Tags:

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] separate the codes for GPU stream synchronization and CPU thread synchronization"


code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.

In a longer term, we would like to give user the flexibility of bypassing the watchdog thread and handle the collective error by themselves.

Test Plan:
python test/distributed/test_c10d_nccl.py NcclErrorHandlingTest


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] separate the codes for GPU stream synchronization and CPU thread synchronization"


code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.

In a longer term, we would like to give user the flexibility of bypassing the watchdog thread and handle the collective error by themselves.

Test Plan:
python test/distributed/test_c10d_nccl.py NcclErrorHandlingTest


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] separate the codes for GPU stream synchronization and CPU thread synchronization"


code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.

In a longer term, we would like to give user the flexibility of bypassing the watchdog thread and handle the collective error by themselves.

Test Plan:
python test/distributed/test_c10d_nccl.py NcclErrorHandlingTest


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -692,42 +722,9 @@ void ProcessGroupNCCL::WorkNCCL::synchronizeInternal(
-    currentStream.synchronize();
-  }
-}
+  // syncrhoize() will block the current stream on the NCCL stream
+  synchronize();
```
nit: this change seems to move `synchronizeStream` after blocking wait. 
In 99% of cases, this should be a no-change to code behavior, because the main thread is being hijacked by blocking wait.
I just worry about the 1% chance where another thread my have a handle on the compute stream too -- in case that, it would be nice to call `synchronizeStream` as early as possible.

shuqiangzhang replied:
You are right, the comments lie about the 'no behavior change part'. The behavior change here is only when blockingwait mode is true or timeout is specified, where user is more concerned about blocking the CPU thread till the collective is done. Will monitor if any users is affected by this change, Later we might further remove GPU stream synchronize() for blockingwait mode.

kwen2501 replied:
I am actually not concerned about behavior change, as you mentioned, it only relates to blocking wait mode.
My comment was more on code robustness, as in, would it be generally better to establish stream dependency asap.

shuqiangzhang replied:
I actually put synchronize() before the cpu wait code in an early version. But it breaks the current barrier behavior which requires the CPU loop happens before the stream sync (https://www.internalfb.com/code/fbsource/[4713fa33ea08c1d66083e1bbe6c1fa4af785b424]/fbcode/caffe2/torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp?lines=695). which I am not sure it is safe to change.  I am planning to remove the barrier blocks in the synchronize() method in future PRs and then we can switch the order

kwen2501 commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -640,9 +642,28 @@ void ProcessGroupNCCL::WorkNCCL::handleException(
 
 void ProcessGroupNCCL::WorkNCCL::synchronize() {
-  // Call Synchronize without a timeout. We use this method to avoid adding a
-  // timeout argument to the public synchronize API.
```
nit: can we add back some documentation on the behavior of this function?

Q: is this method being called at any other places? Just in case the `if (barrierTensor_.defined())` block (moved here) would change their expectation.

shuqiangzhang replied:
"if (barrierTensor_.defined()) " block was actually already here in this API before (within synchronizeInternal). Since We removed synchronizeInternal, so we pop up the block here

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2602,22 +2620,22 @@ def test_nccl_errors_blocking_sigterm(self):
+    @requires_nccl_version((2, 4, 0), "Need NCCL 2.4+ for error checking")
+    @skip_if_lt_x_gpu(3)
+    def test_nccl_non_blocking_wait_with_barrier(self):
+        prev_nccl_async_error_handling = os.environ.get(
+            "TORCH_NCCL_ASYNC_ERROR_HANDLING", None
         )
```
nit: this reminds me of a need for context manager :)

shuqiangzhang replied:
I saw we had decorators to handle envs, e.g.,  @with_nccl_blocking_wait

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2554,6 +2554,24 @@ def _test_nccl_errors_blocking(self, func):
             del process_group
             func()
 
+    def _test_barrier_error(self):
```
nit: if this method do not have more than 1 call site, we can consider flatten it.

shuqiangzhang replied:
put it as a private method as it is shared by both  test_nccl_blocking_wait_with_barrier and test_nccl_nonblocking_wait_with_barrier

kwen2501 commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -2602,22 +2620,22 @@ def test_nccl_errors_blocking_sigterm(self):
+    @requires_nccl()
+    @requires_nccl_version((2, 4, 0), "Need NCCL 2.4+ for error checking")
+    @skip_if_lt_x_gpu(3)
+    def test_nccl_non_blocking_wait_with_barrier(self):
```
nit: do you mind adding comment for the purpose of this test? Thanks!

kwen2501 reviewed (approved):
Giving a tentative approval. 
Can you:
- address the placement of `synchronizeStream` (see comment)
- check if there is any customer of `work.synchronize()` (see comment)
?

kwen2501 commented:
> This PR should not change the existing behavior of work.wait()

While that is true, it also introduced a new behavior:
`work.wait(timeout)`

Can you document this behavior?

And in terms of UX, should user call it like this?
```
try:
  work.wait(timeout)
except:
  # some handling
```

fduwjj commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -404,9 +404,6 @@ class TORCH_API ProcessGroupNCCL : public Backend {
 
    private:
-    // Helper function for synchronize
-    void synchronizeInternal(std::chrono::milliseconds timeout);
```
Nice, this function name is once very confusing to folks.

shuqiangzhang committed:
Update on "[c10d] separate the codes for GPU stream synchronization and CPU thread synchronization"


code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.

In a longer term, we would like to give user the flexibility of bypassing the watchdog thread and handle the collective error by themselves.

Test Plan:
python test/distributed/test_c10d_nccl.py NcclErrorHandlingTest


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] separate the codes for GPU stream synchronization and CPU thread synchronization"


code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.

In a longer term, we would like to give user the flexibility of bypassing the watchdog thread and handle the collective error by themselves.

Test Plan:
python test/distributed/test_c10d_nccl.py NcclErrorHandlingTest


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] separate the codes for GPU stream synchronization and CPU thread synchronization"


code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.

In a longer term, we would like to give user the flexibility of bypassing the watchdog thread and handle the collective error by themselves.

Test Plan:
python test/distributed/test_c10d_nccl.py NcclErrorHandlingTest


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang committed:
Update on "[c10d] separate the codes for GPU stream synchronization and CPU thread synchronization"


code
Summary:
This PR should not change the existing behavior of work.wait(), just
separate the stream synchronization code from the CPU busy wait code.

Also, remove the need of a private synchronization function.

In a longer term, we would like to give user the flexibility of bypassing the watchdog thread and handle the collective error by themselves.

Test Plan:
python test/distributed/test_c10d_nccl.py NcclErrorHandlingTest


cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

shuqiangzhang commented:
@pytorchbot merge
